### PR TITLE
Add timestamp of last successful attestation to verifier API (enhancement #74) and address some missing documentation

### PR DESCRIPTION
**Pull request by stringlytyped**
This pull request fulfills https://github.com/keylime/enhancements/issues/74 by adding a new database column (`last_successful_quote`) and exposing it through the verifier's REST API. Many thanks to @maugustosilva for his help with this [in the Slack channel](https://cloud-native.slack.com/archives/C01ARE2QUTZ/p1676370102354909).

Additionally, I've adding some descriptions to the API documentation for the `last_successful_quote` field and a couple other fields that were missing documentation (namely, `attestation_count` and `last_received_quote`).

See commit messages for further details.

### CODE REVIEW

**Comment by THS-on** [comment]
@stringlytyped thank you for the PR and welcome to Keylime! The changes in general LGTM, but I would name the option `last_successful_attestation` to make it more clear that it refers to the last successful attestation round.

File: docs/rest_apis.rst
Comment (aplanas) [inline docs/rest_apis.rst:122]: Why not to consolidate `null` into 0?

File: keylime/migrations/versions/f838d3cdeead_add_last_successful_quote.py
Comment (aplanas) [inline keylime/migrations/versions/f838d3cdeead_add_last_successful_quote.py:36]: Maybe I am wrong, but verifier_db says that it should not be nullable, I wonder if we can drop this here and put some initial value (0)

Maybe even a similar change can be done for last_received_quote (or not, as at this point can be too late in the migration)

**Review by aplanas — COMMENTED** [review]
Basically LGTM! Just a question about the nullability of the field

File: docs/rest_apis.rst
Comment (maugustosilva) [reply docs/rest_apis.rst:122]: Yeah, in this case, he just followed what I have done originally, and my concern was exactly the migration problems, especially with different SQL backends, especially in case of migrations.

File: keylime/migrations/versions/f838d3cdeead_add_last_successful_quote.py
Comment (maugustosilva) [reply keylime/migrations/versions/f838d3cdeead_add_last_successful_quote.py:36]: So, same problem. I remember an issue with at least one of the backends (MySQL, if memory serves) that lead me, after a quick chat with @mpeters to adopt the more laissez faire policy of set `nullable` to `True` and not setting a value there. I am not going to strongly advocate for it, but if we want to change, we might have to add in a migration for the other column too.

**Comment by maugustosilva** [comment]
@stringlytyped very welcome addition, especially the API documentation (which I forgot to add for `attestation_count` and `last_successful_quote`). 

We are very close to a merge, but please change the column name to `last_successful_attestation` (@THS-on is correct on pointing out that this will avoid the eventual confusion about wether the `quote` was simply received by the `verifier` or was actually fully attested).

As for the `pylint` errors, you can either run `make check`, `./scripts/check_codestyle.sh` (for `pylint` only) or configure the per-commit hooks (which I recommend in case you plan to contribute more)

**Comment by stringlytyped** [comment]
Thanks @THS-on, @aplanas and @maugustosilva for your feedback. It has taken me a moment to get your suggested changes implemented and to fix those formatting issues because I was struggling to get the formatters/linters/etc. working in my local environment. But I think that should be all good to go now, pending CI actions.

File: docs/rest_apis.rst
Comment (stringlytyped) [reply docs/rest_apis.rst]: That's right, I was just following the existing pattern. It does seem to me that there is a slight difference in semantics between "there has never been a successful attestation for this agent" (which results in `last_successful_attestation = 0`) and "there may have been a successful attestation for this agent but there is currently no information about when it occurred" (as would be the case after a keylime install is upgraded and the migration is run for the first time, resulting in  `last_successful_attestation = null`). But I'm not sure if this matters.

**Comment by stringlytyped** [comment]
The only outstanding issue is on fedora-rawhide-x86_64. Looking at the logs, I cannot see how the failures could be due to my changes. Is there a chance this is related to [this bug](https://bugzilla.redhat.com/show_bug.cgi?id=2171376) and the upgrade in Fedora 38 of tpm2-tss and tpm2-tools to versions 4.0.1 and 5.4 respectively?

**Review by aplanas — APPROVED** [review]

**Comment by THS-on** [comment]
> The only outstanding issue is on fedora-rawhide-x86_64. Looking at the logs, I cannot see how the failures could be due to my changes. Is there a chance this is related to [this bug](https://bugzilla.redhat.com/show_bug.cgi?id=2171376) and the upgrade in Fedora 38 of tpm2-tss and tpm2-tools to versions 4.0.1 and 5.4 respectively?

@stringlytyped  yes that failure is in the agent TPM code as you correctly identified and has nothing to do with this change.

Can you integrate the last fixup commit into the two first ones? After that this it should be ready to merge:)

**Comment by stringlytyped** [comment]
> Can you integrate the last fixup commit into the two first ones? After that this it should be ready to merge:)

@THS-on Sure thing. Is it okay if I squash the commits (easier than rebasing manually) or would you prefer the changes to the code base in a separate commit from the updated docs?

**Comment by THS-on** [comment]
@stringlytyped either one is fine for me because all the changes belong logical together.

**Review by maugustosilva — APPROVED** [review]

**Review by THS-on — APPROVED** [review]

**Comment by maugustosilva** [comment]
As we discussed, the Fedora Rawhide failure is not related to your commit. Thank you again, will merge.

**Pull request merged**
